src/hooks/use-form.js

```diff
@@ -0,0 +1,362 @@
+  switch (action.type) {
+    case actionTypes.BLUR:
+    case actionTypes.SET_FIELD_VALUE:
+    case actionTypes.SUBMIT_START:
+      return validate(jsonSchema, values);
+
+    case actionTypes.RESET:
+      return null;
+
+    default:
+      return state;
+  }
+}
+
+/**
+ * Submit status reducer.
+ */
+
+function submitStatusReducer(state, action) {
+  switch (action.type) {
+    case actionTypes.SUBMIT_START:
+      return 'submitStart';
+
+    case actionTypes.SUBMITTING:
+      return 'submitting';
+
+    case actionTypes.SUBMIT_END:
+    case actionTypes.SUBMIT_FAILED:
+      return 'idle';
+
+    default:
+      return state;
+  }
+}
+
+/**
+ * Is submitting reducer.
+ */
+
+function isSubmittingReducer(state, action) {
+  switch (action.type) {
+    case actionTypes.SUBMIT_START:
+    case actionTypes.SUBMITTING:
+      return true;
+
+    case actionTypes.SUBMIT_END:
+    case actionTypes.SUBMIT_FAILED:
+      return false;
+
+    default:
+      return state;
+  }
+}
+
+/**
+ * Form reducer.
+ */
+
+const formReducer = jsonSchema => (state: FormState, action: Action) => {
+  const values = valuesReducer(state.values, action);
+  const meta = metaReducer(state.meta, action);
+  const errors = errorsReducer(state.errors, action, jsonSchema, values);
+  const submitStatus = submitStatusReducer(state.submitStatus, action);
+  const isSubmitting = isSubmittingReducer(state.isSubmitting, action);
+
+  return {
+    errors,
+    isSubmitting,
+    meta,
+    submitStatus,
+    values
+  };
+};
+
+/**
+ * `Options` type.
+ */
+
+type Options = {
+  initialValues: Object,
+  jsonSchema: Object,
+  onSubmit: Submit,
+  onValuesChanged?: (formState: Object) => void
+};
+
+/**
+ * Export `useForm`.
+ */
+
+export default function useForm(options: Options) {
+  const { initialValues = {}, jsonSchema, onSubmit, onValuesChanged } = options;
+  const [state, dispatch] = useReducer(
+    formReducer(jsonSchema),
+    {
+      errors: null,
+      isSubmitting: false,
+      meta: {},
+      submitStatus: 'idle',
+      values: initialValues
+    }
+  );
+
+  const setFieldValue = useCallback((field: string, value: any) => {
+    dispatch({
+      payload: { field, value },
+      type: actionTypes.SET_FIELD_VALUE
+    });
+  }, []);
+
+  const blurField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.BLUR
+    });
+  }, []);
+
+  const focusField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.FOCUS
+    });
+  }, []);
+
+  const registerField = useCallback((field: string) => {
+    dispatch({
+      payload: { field },
+      type: actionTypes.REGISTER_FIELD
+    });
+  }, []);
+
+  const reset = useCallback(() => {
+    dispatch({
+      payload: { initialValues },
+      type: actionTypes.RESET
+    });
+  }, [initialValues]);
+
+  const submit = useCallback((event: ?SyntheticInputEvent<any>) => {
+    if (event && event.preventDefault) {
```

Comment:
According to [SyntheticInputEvent type](https://github.com/facebook/flow/blob/v0.96.1/lib/react-dom.js#L133), if it exists it will always have a `preventDefault` method. Validating that `preventDefault` exists, doesn't validate that it's a function, invoking it will throw an error.